File: window/js/GfxUtil.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-20  Jed Young  Creation
//

fan.vaseWindow.GfxUtil = {}

fan.vaseWindow.GfxUtil.fontToCss = function(font)
{
  var s = "";
  if (font.m_bold)   s += "bold ";
  if (font.m_italic) s += "italic ";
  s += font.m_size + "px ";
  s += font.m_$name;
  return s;
}

fan.vaseWindow.GfxUtil.uriToImageSrc = function(uri)
{
  if ("fan" == uri.scheme())
    return "/pod/" + uri.host() + uri.pathStr()
  else
    return uri.toStr();
}

fan.vaseWindow.GfxUtil.addEventListener = function(obj, type, func)
{
  // for IE
  if (!obj.addEventListener) {
    obj.attachEvent("on" + type, func);
  }
  else {
    if (type == "mousewheel") {
      var passiveSupported = false;
      try {
          var options = Object.defineProperty({}, "passive", {
              get: function() {
                  passiveSupported = true;
              }
          });
          window.addEventListener("test", null, options);
      } catch(err) {}

      try {
        obj.addEventListener("mousewheel", func, passiveSupported ? { passive: false } : false)
      }
      catch(err) {
        obj.addEventListener("DOMMouseScroll", func, false);
        window.onmousewheel = document.onmousewheel = func;
      }
    }
    else {
      obj.addEventListener(type, func, false);
    }
  }
}

fan.vaseWindow.GfxUtil.doJsPath = function(cx, path)
{
  var size = path.steps().size();
  cx.beginPath();
  for (var i =0; i < size; ++i)
  {
    var s = path.steps().get(i);

    if (s instanceof fan.vaseGraphics.PathMoveTo)
    {
      cx.moveTo(s.m_x, s.m_y);
    }
    else if (s instanceof fan.vaseGraphics.PathLineTo)
    {
      cx.lineTo(s.m_x, s.m_y);
    }
    else if (s instanceof fan.vaseGraphics.PathQuadTo)
    {
      cx.quadraticCurveTo(s.m_cx, s.m_cy, s.m_x, s.m_y);
    }
    else if (s instanceof fan.vaseGraphics.PathCubicTo)
    {
      cx.bezierCurveTo(s.m_cx1, s.m_cy1, s.m_cx2, s.m_cy2, s.m_x, s.m_y);
    }
    else if (s instanceof fan.vaseGraphics.PathClose)
    {
      cx.closePath();
    }
    else if (s instanceof fan.vaseGraphics.PathArc) {
      var sa  = Math.PI / 180 * s.m_startAngle;
      var ea  = Math.PI / 180 * (s.m_startAngle + s.m_arcAngle);
      cx.arc(s.m_cx, s.m_cy, s.m_radius, sa, ea, true);
    }
    else
    {
      throw fan.sys.Err.make("unreachable");
    }
  }
}

fan.vaseWindow.GfxUtil.doJsTransform = function(cx, trans)
{
  cx.transform(
       trans.m_a,
       trans.m_b,
       trans.m_c,
       trans.m_d,
       trans.m_e,
       trans.m_f
     );
}

File: window/java/fan/vaseWindow/GfxUtil.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-09-18  Jed Young  Creation
//
package fan.vaseWindow;

import fan.vaseGraphics.Point;

public class GfxUtil {
  public static float[] toFloats(fan.sys.List ints) {
    int size = (int) ints.size();
    float[] a = new float[size];
    for (int i = 0; i < size; i += 1) {
      long p = (Long) ints.get(i);
      a[i] = (float) p;
    }
    return a;
  }

  public static int[] toIntsX(fan.sys.List points) {
    int size = (int) points.size();
    int[] a = new int[size];
    for (int i = 0; i < size; i++) {
      Point p = (Point) points.get(i);
      a[i] = (int) p.x;
    }
    return a;
  }

  public static int[] toIntsY(fan.sys.List points) {
    int size = (int) points.size();
    int[] a = new int[size];
    for (int i = 0; i < size; i++) {
      Point p = (Point) points.get(i);
      a[i] = (int) p.y;
    }
    return a;
  }

  public static float[] intsToFloats(int[] a) {
    float[] b = new float[a.length];
    for (int i = 0; i < a.length; ++i) {
      b[i] = a[i];
    }
    return b;
  }
}